src/thread/n3ds/SDL_syssem.c

```diff
@@ -57,44 +58,50 @@ SDL_CreateSemaphore(Uint32 initial_value)
 */
 void SDL_DestroySemaphore(SDL_sem *sem)
 {
-    if (sem) {
-        SDL_free(sem);
-    }
+    SDL_free(sem);
 }
 
 int SDL_SemWaitTimeoutNS(SDL_sem *sem, Sint64 timeoutNS)
 {
-    int retval;
-
     if (sem == NULL) {
         return SDL_InvalidParamError("sem");
     }
 
     if (timeoutNS == SDL_MUTEX_MAXWAIT) {
         LightSemaphore_Acquire(&sem->semaphore, 1);
-        retval = 0;
-    } else {
-        int return_code = LightSemaphore_TryAcquire(&sem->semaphore, 1);
-        if (return_code != 0) {
-            /* FIXME: Does this code guarantee a wall clock timeout here?
-             *        Can we handle sub-millisecond delays? */
-            u32 timeout = (u32)SDL_NS_TO_MS(timeoutNS);
-            for (u32 i = 0; i < timeout; i++) {
-                svcSleepThread(1000000LL);
-                return_code = LightSemaphore_TryAcquire(&sem->semaphore, 1);
-                if (return_code == 0) {
-                    break;
-                }
-            }
-        }
-        retval = return_code != 0 ? SDL_MUTEX_TIMEDOUT : 0;
+        return 0;
+    }
+
+    if (timeoutNS == 0) {
+        int result = LightSemaphore_TryAcquire(&sem->semaphore, 1);
+        /* TryWait will block unless we wait about 50 microseconds */
+        SDL_DelayNS(SDL_US_TO_NS(50));
```

Comment:
This seems wrong. Why would you need to delay after the tryacquire and before the code that acquired it?

Reply:
No, something else is happening in the system that is requiring the wait. If you have nothing else operating on the semaphore, do you still need the wait?

Reply:
I've spent the last few hours trying to debug the issue, here are my findings:
- When just creating a semaphore with a high count and calling TryWait as fast as possible to get it down to 0, it works fine without any delay.
- As soon as the consumer (calling TryWait) and the producer (calling Post) are on different threads, there is a chance of blocking.

To expand further on that last point:
- It seems that any newly created thread will starve its parent unless it goes to sleep
- With the default settings of `testsem` (10 threads), it takes about 50µs for all the consumer threads to be asleep and the main thread (producer) to be scheduled.
- When bringing it down to one consumer thread, then any sleep duration in the thread will allow the main thread to be scheduled.
- It seems that even when the two threads have the same affinity, the scheduler will not change threads and whichever one is currently running will starve all the others.

Considering all this, what should be done about TryWait ? Should it keep the current 50µs delay to pass `testsem` ? As soon as there are 2 consumer threads, it already takes a delay of about 15µs to avoid starvation.

Reply:
Maybe what we need to do instead is have whoever is releasing the semaphore sleep immediately afterwards?

Reply:
> This seems like a bug in the N3DS runtime or scheduler. I don't think we should fake things to get the tests to pass, as that'll bite someone later who is relying on these functions. Does the N3DS not have pre-emptive multitasking?

From what I've found, the OS uses cooperative multitasking instead of preemptive. Which would explain why there are no context switching from a child thread unless it sleeps.

> Maybe what we need to do instead is have whoever is releasing the semaphore sleep immediately afterwards?

This did not work, on its own. But here are a few things I noticed:
- in testsem, the main thread does a busy wait until the semaphore is back to 0, to avoid starvation, it should have a wait:
  ```c
  while (SDL_SemValue(sem)) {
    SDL_DelayNS(1);
  }
  ```
- in TryWait, a delay of 1ns avoids starvation and lets all threads consume the semaphore equally.
- in Post, a delay of 1ns will make the first consumer thread starve all the other threads from getting a successful acquire.

Reply:
Making the code more friendly to cooperative multitasking seems like a good idea. Feel free to submit a new PR (or change this one) to do that.